sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/Counter.java

```diff
@@ -1084,6 +1153,7 @@ public Integer getAggregate() {
 
     @Override
     public Counter<Integer> merge(Counter<Integer> that) {
+      setDirty();
```

Comment:
I moved setDirty() to finally block for a race condition:
for example, if the updater called commiting() and committed() between setDirty() and the counter is mutated. The dirty state is lost.

There might be cases that the dirty bit is set unnecessarily. But, I think it is fine as long as the implementation is simple and easy to reasoning the correctness "dirty bit never get lost"